SagaTestFixture and AggregateTestFixture throw exceptions when accessing java.lang.* classes on Java 17

<!-- Please use markdown (https://guides.github.com/features/mastering-markdown/) semantics throughout the bug description. -->

### Basic information

* Axon Framework version: 4.5.9
* JDK version: OpenJDK 64-Bit Server VM Temurin-17.0.2+8 (build 17.0.2+8, mixed mode, sharing)
* Complete executable reproducer if available (e.g. GitHub Repo): Test snippets included below

### SagaTestFixture

#### Steps to reproduce

The deadline manager allows an arbitrary object for a payload, including types from the `java.lang` package like `String`. 
`SagaTestFixture` allows asserting deadline manager's payloads and achieves it through the reflection mechanism for the user's convenience. 
For these reasons, the `EqualFieldsMatcher` may inspect the `java.lang.String`, and make its internal fields accessible. 
Java 17 denies such access by default since [JEP 403](https://openjdk.java.net/jeps/403) and throws an exception.
The issue can be worked around by opening the `java.lang` package to the unnamed module: `--add-opens java.base/java.lang=ALL-UNNAMED`.

Reproducer:
```
class SomeSagaTest {

    private final SagaTestFixture<SomeSaga> testee = new SagaTestFixture(SomeSaga.class);

    @Test
    void matchesStringDeadlinePayload() {
        testee.givenAPublished(new SomeEvent("id"))
              .whenTimeElapses(Duration.ofMinutes(1))
              .expectDeadlinesMet("string payload");
    }

    @Saga
    public static class SomeSaga {

        @Autowired
        private transient DeadlineManager deadlineManager;

        @StartSaga
        @SagaEventHandler(associationProperty = "id")
        public void on(SomeEvent event) {
            deadlineManager.schedule(Duration.ofSeconds(10), "some-deadline", "string payload");
        }
    }

    public static class SomeEvent {

        private final String id;

        public SomeEvent(String id) {
            this.id = id;
        }

        public String getId() {
            return id;
        }
    }
}
```

#### Expected behaviour
Test passes.

#### Actual behaviour
The exception is thrown:
```
java.lang.reflect.InaccessibleObjectException: Unable to make field private final byte[] java.lang.String.value accessible: module java.base does not "opens java.lang" to unnamed module @18e8568

	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
	at java.base/java.lang.reflect.Field.checkCanSetAccessible(Field.java:178)
	at java.base/java.lang.reflect.Field.setAccessible(Field.java:172)
	at org.axonframework.test.matchers.EqualFieldsMatcher.fieldsMatch(EqualFieldsMatcher.java:78)
	at org.axonframework.test.matchers.EqualFieldsMatcher.matchesSafely(EqualFieldsMatcher.java:71)
	at org.axonframework.test.matchers.EqualFieldsMatcher.matches(EqualFieldsMatcher.java:66)
	at org.axonframework.test.matchers.ExactSequenceMatcher.matchesList(ExactSequenceMatcher.java:56)
	at org.axonframework.test.matchers.ListMatcher.matches(ListMatcher.java:52)
	at org.axonframework.test.matchers.PayloadsMatcher.matches(PayloadsMatcher.java:58)
	at org.axonframework.test.deadline.DeadlineManagerValidator.assertDeadlinesMetMatching(DeadlineManagerValidator.java:132)
	at org.axonframework.test.deadline.DeadlineManagerValidator.assertDeadlinesMet(DeadlineManagerValidator.java:156)
	at org.axonframework.test.saga.FixtureExecutionResultImpl.expectDeadlinesMet(FixtureExecutionResultImpl.java:359)
	at com.example.sagatest.SomeSagaTest.matchesStringDeadlinePayload(SomeSagaTest.java:21) <31 internal lines>
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511) <9 internal lines>
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511) <25 internal lines>
```

### SagaTestFixture

#### Steps to reproduce

An aggregate allows an arbitrary object for a field, including types from the `java.lang` package like `Class`.
`AggregateTestFixture` detects illegal state changes and achieves it through the reflection mechanism.
While doing this it may inspect the `java.lang.Class`, and make its internal fields accessible.
Java 17 denies such access by default since [JEP 403](https://openjdk.java.net/jeps/403) and throws an exception.
The issue can be worked around by opening the `java.lang` package to the unnamed module: `--add-opens java.base/java.lang=ALL-UNNAMED`.

Reproducer:
```
public class SomeAggregateTest {

    private final AggregateTestFixture<SomeAggregate> testee = new AggregateTestFixture<>(SomeAggregate.class);

    @Test
    void publishesSomeEventOnSomeCommand() {
        testee.givenNoPriorActivity()
              .when(new SomeCommand("id"))
              .expectEvents(new SomeEvent("id"));
    }

    @Aggregate
    public static class SomeAggregate {

        @AggregateIdentifier
        private String id;
        private Class<String> rawClassField = String.class;

        public SomeAggregate() {
        }

        @CommandHandler
        public SomeAggregate(SomeCommand command) {
            AggregateLifecycle.apply(new SomeEvent(command.getId()));
        }

        @EventSourcingHandler
        public void on(SomeEvent event) {
            id = event.getId();
        }
    }

    public static class SomeCommand {

        @TargetAggregateIdentifier
        private final String id;

        public SomeCommand(String id) {
            this.id = id;
        }

        public String getId() {
            return id;
        }
    }


    public static class SomeEvent {

        private final String id;

        public SomeEvent(String id) {
            this.id = id;
        }

        public String getId() {
            return id;
        }
    }
}
```

#### Expected behaviour
Test passes.

#### Actual behaviour
The exception is thrown:
```
org.axonframework.test.FixtureExecutionException: An Exception occurred while reconstructing the Aggregate from given and published events. This may be an indication that the aggregate cannot be recreated from its events.

	at org.axonframework.test.aggregate.AggregateTestFixture.detectIllegalStateChanges(AggregateTestFixture.java:566)
	at org.axonframework.test.aggregate.AggregateTestFixture.when(AggregateTestFixture.java:455)
	at org.axonframework.test.aggregate.AggregateTestFixture.when(AggregateTestFixture.java:438)
	at com.example.sagatest.SomeAggregateTest.publishesSomeEventOnSomeCommand(SomeAggregateTest.java:19) <31 internal lines>
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511) <9 internal lines>
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511) <25 internal lines>
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make field private final java.lang.Class java.lang.Class.componentType accessible: module java.base does not "opens java.lang" to unnamed module @18e8568
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
	at java.base/java.lang.reflect.Field.checkCanSetAccessible(Field.java:178)
	at java.base/java.lang.reflect.Field.setAccessible(Field.java:172)
	at org.axonframework.common.MemberAccessibilityCallback.run(MemberAccessibilityCallback.java:43)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:318)
	at org.axonframework.common.ReflectionUtils.ensureAccessible(ReflectionUtils.java:176)
	at org.axonframework.test.aggregate.AggregateTestFixture.ensureValuesEqual(AggregateTestFixture.java:605)
	at org.axonframework.test.aggregate.AggregateTestFixture.ensureValuesEqual(AggregateTestFixture.java:610)
	at org.axonframework.test.aggregate.AggregateTestFixture.assertValidWorkingAggregateState(AggregateTestFixture.java:587)
	at org.axonframework.test.aggregate.AggregateTestFixture.detectIllegalStateChanges(AggregateTestFixture.java:558)
	... 70 more
```

--- a/messaging/src/main/java/org/axonframework/common/ReflectionUtils.java
+++ b/messaging/src/main/java/org/axonframework/common/ReflectionUtils.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2010-2020. Axon Framework
+ * Copyright (c) 2010-2022. Axon Framework
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -492,6 +492,25 @@ public abstract class ReflectionUtils {
         throw new IllegalStateException(
                 String.format(UNSUPPORTED_MEMBER_TYPE_EXCEPTION_MESSAGE, member.getClass().getName())
         );
+    }
+
+    /**
+     * Check whether the given Object is a primitive types. Validates both boxed and unboxed instances.
+     *
+     * @param o The Object to check whether it's a primitive type.
+     * @return {@code true} if the given Object is a primitive, {@code false} otherwise.
+     */
+    public static boolean isPrimitive(Object o) {
+        return o.getClass().isPrimitive()
+                || o.getClass().isAssignableFrom(Boolean.class)
+                || o.getClass().isAssignableFrom(Byte.class)
+                || o.getClass().isAssignableFrom(Character.class)
+                || o.getClass().isAssignableFrom(Short.class)
+                || o.getClass().isAssignableFrom(Integer.class)
+                || o.getClass().isAssignableFrom(Float.class)
+                || o.getClass().isAssignableFrom(Double.class)
+                || o.getClass().isAssignableFrom(Long.class)
+                || o.getClass().isAssignableFrom(String.class);
     }
 
     private ReflectionUtils() {

--- a/messaging/src/main/java/org/axonframework/common/ReflectionUtils.java
+++ b/messaging/src/main/java/org/axonframework/common/ReflectionUtils.java
@@ -495,7 +495,9 @@ public abstract class ReflectionUtils {
     }
 
     /**
-     * Check whether the given Object is a primitive types. Validates both boxed and unboxed instances.
+     * Check whether the given Object is a primitive type. Validates both boxed and unboxed instances.
+     * <p>
+     * Note that this check includes {@link String} as well.
      *
      * @param o The Object to check whether it's a primitive type.
      * @return {@code true} if the given Object is a primitive, {@code false} otherwise.

--- a/messaging/src/test/java/org/axonframework/common/ReflectionUtilsTest.java
+++ b/messaging/src/test/java/org/axonframework/common/ReflectionUtilsTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2010-2020. Axon Framework
+ * Copyright (c) 2010-2022. Axon Framework
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -292,6 +292,55 @@ class ReflectionUtilsTest {
         Constructor<SomeTypeWithMethods> constructor = SomeTypeWithMethods.class.getDeclaredConstructor();
         String memberGenericString = getMemberGenericString(constructor);
         assertEquals(constructor.toGenericString(), memberGenericString);
+    }
+
+    @Test
+    void testIsPrimitive() {
+        boolean bool = true;
+        assertTrue(ReflectionUtils.isPrimitive(bool));
+        Boolean boxedBool = true;
+        assertTrue(ReflectionUtils.isPrimitive(boxedBool));
+
+        byte b = (byte) 42;
+        assertTrue(ReflectionUtils.isPrimitive(b));
+        Byte boxedByte = (byte) 42;
+        assertTrue(ReflectionUtils.isPrimitive(boxedByte));
+
+        char c = 'a';
+        assertTrue(ReflectionUtils.isPrimitive(c));
+        Character boxedChar = 'a';
+        assertTrue(ReflectionUtils.isPrimitive(boxedChar));
+
+        short s = (short) 42;
+        assertTrue(ReflectionUtils.isPrimitive(s));
+        Short boxedShort = (short) 42;
+        assertTrue(ReflectionUtils.isPrimitive(boxedShort));
+
+        int i = 42;
+        assertTrue(ReflectionUtils.isPrimitive(i));
+        Integer boxedInteger = 42;
+        assertTrue(ReflectionUtils.isPrimitive(boxedInteger));
+
+        float f = 42F;
+        assertTrue(ReflectionUtils.isPrimitive(f));
+        Float boxedFloat = 42F;
+        assertTrue(ReflectionUtils.isPrimitive(boxedFloat));
+
+        double d = 42.42;
+        assertTrue(ReflectionUtils.isPrimitive(d));
+        Double boxedDouble = 42.42;
+        assertTrue(ReflectionUtils.isPrimitive(boxedDouble));
+
+        long l = 42L;
+        assertTrue(ReflectionUtils.isPrimitive(l));
+        Long boxedLong = 42L;
+        assertTrue(ReflectionUtils.isPrimitive(boxedLong));
+
+        String sampleString = "some-text";
+        assertTrue(ReflectionUtils.isPrimitive(sampleString));
+
+        SomeType nonPrimitive = new SomeType();
+        assertFalse(ReflectionUtils.isPrimitive(nonPrimitive));
     }
 
     @SuppressWarnings("FieldCanBeLocal")

--- a/test/src/main/java/org/axonframework/test/aggregate/Reporter.java
+++ b/test/src/main/java/org/axonframework/test/aggregate/Reporter.java
@@ -1,11 +1,11 @@
 /*
- * Copyright (c) 2010-2014. Axon Framework
+ * Copyright (c) 2010-2022. Axon Framework
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
- *     http://www.apache.org/licenses/LICENSE-2.0
+ *    http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
@@ -195,15 +195,35 @@ public class Reporter {
     }
 
     /**
+     * Report an error due to a difference between the primitive message payloads.
+     *
+     * @param messageType The (runtime) type of the message the difference was found in.
+     * @param actual      The actual value of the payload.
+     * @param expected    The expected value of the payload.
+     */
+    public void reportDifferentPrimitivePayloads(Class<?> messageType, Object actual, Object expected) {
+        throw new AxonAssertionError("One of the messages contained a different payload than expected"
+                                             + NEWLINE + NEWLINE
+                                             + "The message of type [" + messageType.getSimpleName() + "], "
+                                             + "was not as expected."
+                                             + NEWLINE
+                                             + "Expected <" //NOSONAR
+                                             + nullSafeToString(expected)
+                                             + "> but got <"
+                                             + nullSafeToString(actual)
+                                             + ">"
+                                             + NEWLINE);
+    }
+
+    /**
      * Report an error due to a difference in one of the fields of the message payload.
      *
-     * @param messageType The (runtime) type of the message the difference was found in
-     * @param field       The field that contains the difference
-     * @param actual      The actual value of the field
-     * @param expected    The expected value of the field
+     * @param messageType The (runtime) type of the message the difference was found in.
+     * @param field       The field that contains the difference.
+     * @param actual      The actual value of the field.
+     * @param expected    The expected value of the field.
      */
-    public void reportDifferentPayloads(Class<?> messageType, Field field, Object actual,
-                                        Object expected) {
+    public void reportDifferentPayloads(Class<?> messageType, Field field, Object actual, Object expected) {
         StringBuilder sb = new StringBuilder("One of the messages contained different values than expected");
         sb.append(NEWLINE)
           .append(NEWLINE)

--- a/test/src/main/java/org/axonframework/test/aggregate/ResultValidatorImpl.java
+++ b/test/src/main/java/org/axonframework/test/aggregate/ResultValidatorImpl.java
@@ -30,6 +30,7 @@ import org.axonframework.test.deadline.StubDeadlineManager;
 import org.axonframework.test.matchers.EqualFieldsMatcher;
 import org.axonframework.test.matchers.FieldFilter;
 import org.axonframework.test.matchers.MapEntryMatcher;
+import org.axonframework.test.matchers.Matchers;
 import org.axonframework.test.matchers.PayloadMatcher;
 import org.hamcrest.CoreMatchers;
 import org.hamcrest.Matcher;
@@ -442,12 +443,18 @@ public class ResultValidatorImpl<T> implements ResultValidator<T>, CommandCallba
         if (!expectedPayload.getClass().equals(actualPayload.getClass())) {
             return false;
         }
-        EqualFieldsMatcher<Object> matcher = new EqualFieldsMatcher<>(expectedPayload, fieldFilter);
+        EqualFieldsMatcher<Object> matcher = Matchers.equalTo(expectedPayload, fieldFilter);
         if (!matcher.matches(actualPayload)) {
-            reporter.reportDifferentPayloads(expectedPayload.getClass(),
-                                             matcher.getFailedField(),
-                                             matcher.getFailedFieldActualValue(),
-                                             matcher.getFailedFieldExpectedValue());
+            if (matcher.isFailedPrimitive()) {
+                reporter.reportDifferentPrimitivePayloads(expectedPayload.getClass(),
+                                                          matcher.getFailedFieldActualValue(),
+                                                          matcher.getFailedFieldExpectedValue());
+            } else {
+                reporter.reportDifferentPayloads(expectedPayload.getClass(),
+                                                 matcher.getFailedField(),
+                                                 matcher.getFailedFieldActualValue(),
+                                                 matcher.getFailedFieldExpectedValue());
+            }
         }
         return true;
     }

--- a/test/src/main/java/org/axonframework/test/matchers/EqualFieldsMatcher.java
+++ b/test/src/main/java/org/axonframework/test/matchers/EqualFieldsMatcher.java
@@ -1,11 +1,11 @@
 /*
- * Copyright (c) 2010-2014. Axon Framework
+ * Copyright (c) 2010-2022. Axon Framework
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
- *     http://www.apache.org/licenses/LICENSE-2.0
+ *    http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
@@ -22,6 +22,8 @@ import org.hamcrest.Description;
 import java.lang.reflect.Field;
 import java.util.Objects;
 
+import static org.axonframework.common.ReflectionUtils.isPrimitive;
+
 /**
  * Matcher that will match an Object if all the fields on that Object contain values equal to the same field in the
  * expected instance.
@@ -37,10 +39,11 @@ public class EqualFieldsMatcher<T> extends BaseMatcher<T> {
     private Field failedField;
     private Object failedFieldExpectedValue;
     private Object failedFieldActualValue;
+    private boolean failedPrimitive;
 
     /**
-     * Initializes an EqualFieldsMatcher that will match an object with equal properties as the given
-     * {@code expected} object.
+     * Initializes an EqualFieldsMatcher that will match an object with equal properties as the given {@code expected}
+     * object.
      *
      * @param expected The expected object
      */
@@ -49,8 +52,8 @@ public class EqualFieldsMatcher<T> extends BaseMatcher<T> {
     }
 
     /**
-     * Initializes an EqualFieldsMatcher that will match an object with equal properties as the given
-     * {@code expected} object.
+     * Initializes an EqualFieldsMatcher that will match an object with equal properties as the given {@code expected}
+     * object.
      *
      * @param expected The expected object
      * @param filter   The filter describing the fields to include in the comparison
@@ -60,13 +63,22 @@ public class EqualFieldsMatcher<T> extends BaseMatcher<T> {
         this.filter = filter;
     }
 
-    @SuppressWarnings({"unchecked"})
     @Override
-    public boolean matches(Object item) {
-        return expected.getClass().isInstance(item) && matchesSafely(item);
+    public boolean matches(Object actual) {
+        return expected.getClass().isInstance(actual) && matchesSafely(actual);
     }
 
     private boolean matchesSafely(Object actual) {
+        if ((isPrimitive(expected) || isPrimitive(actual))) {
+            if (expected.equals(actual)) {
+                return true;
+            } else {
+                failedPrimitive = true;
+                failedFieldExpectedValue = expected;
+                failedFieldActualValue = actual;
+                return false;
+            }
+        }
         return expected.getClass().equals(actual.getClass())
                 && fieldsMatch(expected.getClass(), expected, actual);
     }
@@ -107,8 +119,8 @@ public class EqualFieldsMatcher<T> extends BaseMatcher<T> {
     }
 
     /**
-     * Returns the expected value of a failed field comparison, if any. This value is only populated after {@link
-     * #matches(Object)} is called and a mismatch has been detected.
+     * Returns the expected value of a failed field comparison, if any. This value is only populated after
+     * {@link #matches(Object)} is called and a mismatch has been detected.
      *
      * @return the expected value of the field that failed comparison, if any
      */
@@ -117,13 +129,22 @@ public class EqualFieldsMatcher<T> extends BaseMatcher<T> {
     }
 
     /**
-     * Returns the actual value of a failed field comparison, if any. This value is only populated after {@link
-     * #matches(Object)} is called and a mismatch has been detected.
+     * Returns the actual value of a failed field comparison, if any. This value is only populated after
+     * {@link #matches(Object)} is called and a mismatch has been detected.
      *
      * @return the actual value of the field that failed comparison, if any
      */
     public Object getFailedFieldActualValue() {
         return failedFieldActualValue;
+    }
+
+    /**
+     * Returns whether this matcher failed matching primitive types.
+     *
+     * @return {@code true} if this matcher failed matching primitive types, {@code false} otherwise.
+     */
+    public boolean isFailedPrimitive() {
+        return failedPrimitive;
     }
 
     @Override

--- a/test/src/main/java/org/axonframework/test/matchers/EqualFieldsMatcher.java
+++ b/test/src/main/java/org/axonframework/test/matchers/EqualFieldsMatcher.java
@@ -140,6 +140,8 @@ public class EqualFieldsMatcher<T> extends BaseMatcher<T> {
 
     /**
      * Returns whether this matcher failed matching primitive types.
+     * <p>
+     * Note that this may include an expected or actual of type {@link String}.
      *
      * @return {@code true} if this matcher failed matching primitive types, {@code false} otherwise.
      */

--- a/test/src/main/java/org/axonframework/test/saga/CommandValidator.java
+++ b/test/src/main/java/org/axonframework/test/saga/CommandValidator.java
@@ -1,11 +1,11 @@
 /*
- * Copyright (c) 2010-2014. Axon Framework
+ * Copyright (c) 2010-2022. Axon Framework
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
- *     http://www.apache.org/licenses/LICENSE-2.0
+ *    http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
@@ -20,6 +20,7 @@ import org.axonframework.commandhandling.CommandMessage;
 import org.axonframework.test.AxonAssertionError;
 import org.axonframework.test.matchers.EqualFieldsMatcher;
 import org.axonframework.test.matchers.FieldFilter;
+import org.axonframework.test.matchers.Matchers;
 import org.axonframework.test.utils.RecordingCommandBus;
 import org.hamcrest.Description;
 import org.hamcrest.Matcher;
@@ -130,15 +131,23 @@ public class CommandValidator {
                                                     expected.getClass().getSimpleName(),
                                                     actual.getClass().getSimpleName()));
             }
-            EqualFieldsMatcher<Object> matcher = new EqualFieldsMatcher<>(expected, fieldFilter);
+            EqualFieldsMatcher<Object> matcher = Matchers.equalTo(expected, fieldFilter);
             if (!matcher.matches(actual)) {
-                throw new AxonAssertionError(format("Unexpected command at index %s (0-based). "
-                                                            + "Field value of '%s.%s', expected <%s>, but got <%s>",
-                                                    commandIndex,
-                                                    expected.getClass().getSimpleName(),
-                                                    matcher.getFailedField().getName(),
-                                                    matcher.getFailedFieldExpectedValue(),
-                                                    matcher.getFailedFieldActualValue()));
+                if (matcher.isFailedPrimitive()) {
+                    throw new AxonAssertionError(format("Unexpected primitive typed command at index %s (0-based). "
+                                                                + "Expected <%s>, but got <%s>",
+                                                        commandIndex,
+                                                        matcher.getFailedFieldExpectedValue(),
+                                                        matcher.getFailedFieldActualValue()));
+                } else {
+                    throw new AxonAssertionError(format("Unexpected command at index %s (0-based). "
+                                                                + "Field value of '%s.%s', expected <%s>, but got <%s>",
+                                                        commandIndex,
+                                                        expected.getClass().getSimpleName(),
+                                                        matcher.getFailedField().getName(),
+                                                        matcher.getFailedFieldExpectedValue(),
+                                                        matcher.getFailedFieldActualValue()));
+                }
             }
         }
     }

--- a/test/src/test/java/org/axonframework/test/aggregate/ResultValidatorImplTest.java
+++ b/test/src/test/java/org/axonframework/test/aggregate/ResultValidatorImplTest.java
@@ -1,11 +1,11 @@
 /*
- * Copyright (c) 2010-2019. Axon Framework
+ * Copyright (c) 2010-2022. Axon Framework
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
- *     http://www.apache.org/licenses/LICENSE-2.0
+ *    http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
@@ -19,15 +19,19 @@ package org.axonframework.test.aggregate;
 import org.axonframework.eventhandling.EventMessage;
 import org.axonframework.test.AxonAssertionError;
 import org.axonframework.test.matchers.MatchAllFieldFilter;
-import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.*;
 
 import java.util.List;
 
 import static java.util.Collections.*;
 import static org.axonframework.eventhandling.GenericEventMessage.asEventMessage;
-import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.*;
 
+/**
+ * Test class validating the {@link ResultValidatorImpl}.
+ *
+ * @author bliessens
+ */
 class ResultValidatorImplTest {
 
     private ResultValidator<?> validator = new ResultValidatorImpl<>(actualEvents(),
@@ -50,7 +54,7 @@ class ResultValidatorImplTest {
     }
 
     @Test
-    void shouldSuccesfullyCompareEqualMetadata() {
+    void shouldSuccessfullyCompareEqualMetadata() {
         EventMessage<?> expected = actualEvents().iterator().next().andMetaData(singletonMap("key1", "value1"));
 
         validator.expectEvents(expected);
@@ -66,15 +70,34 @@ class ResultValidatorImplTest {
         String s2 = String.valueOf(0);
         assertEquals(s1, s2);
 
-        // the hash code is cached in a String
+        //noinspection unused -> the hash code is cached in a String
         int ignored = s1.hashCode();
 
         validator.expectEvents(s2);
+    }
+
+    @Test
+    void shouldReportFailureForFailedPrimitiveMatching() {
+        validator = new ResultValidatorImpl<>(singletonList(asEventMessage("some-string")),
+                                              new MatchAllFieldFilter(emptyList()),
+                                              () -> null,
+                                              null);
+
+        assertThrows(AxonAssertionError.class, () -> validator.expectEvents("some-other-string"));
+    }
+
+    @Test
+    void shouldReportFailureForFailedFieldMatching() {
+        validator = new ResultValidatorImpl<>(singletonList(asEventMessage(new MyEvent("some-string", 1))),
+                                              new MatchAllFieldFilter(emptyList()),
+                                              () -> null,
+                                              null);
+
+        assertThrows(AxonAssertionError.class, () -> validator.expectEvents(new MyEvent("some-other-string", 1)));
     }
 
     private List<EventMessage<?>> actualEvents() {
         return singletonList(asEventMessage(new MyEvent("aggregateId", 123))
                                      .andMetaData(singletonMap("key1", "value1")));
     }
-
 }

--- a/test/src/test/java/org/axonframework/test/matchers/EqualFieldsMatcherTest.java
+++ b/test/src/test/java/org/axonframework/test/matchers/EqualFieldsMatcherTest.java
@@ -1,11 +1,11 @@
 /*
- * Copyright (c) 2010-2012. Axon Framework
+ * Copyright (c) 2010-2022. Axon Framework
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
- *     http://www.apache.org/licenses/LICENSE-2.0
+ *    http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
@@ -19,19 +19,20 @@ package org.axonframework.test.matchers;
 import org.axonframework.test.aggregate.MyEvent;
 import org.axonframework.test.aggregate.MyOtherEvent;
 import org.hamcrest.StringDescription;
-import org.junit.jupiter.api.BeforeEach;
-import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.*;
 
 import static org.junit.jupiter.api.Assertions.*;
 
 /**
+ * Test class validating the {@link EqualFieldsMatcher}.
+ *
  * @author Allard Buijze
  */
 class EqualFieldsMatcherTest {
 
     private EqualFieldsMatcher<MyEvent> testSubject;
     private MyEvent expectedEvent;
-    private String aggregateId = "AggregateId";
+    private final String aggregateId = "AggregateId";
 
     @BeforeEach
     void setUp() {
@@ -94,5 +95,57 @@ class EqualFieldsMatcherTest {
         StringDescription description = new StringDescription();
         testSubject.describeTo(description);
         assertEquals("org.axonframework.test.aggregate.MyEvent (failed on field 'someValue')", description.toString());
+    }
+
+    /**
+     * This test is introduced to validate whether the matcher doesn't delve into the boxed primitives, since reflective
+     * access is restricted for e.g. JDK 17.
+     */
+    @Test
+    void testMatchesReturnsTrueForPrimitives() {
+        EqualFieldsMatcher<Boolean> booleanMatcher = Matchers.equalTo(false);
+        assertTrue(booleanMatcher.matches(false));
+        assertFalse(booleanMatcher.matches(true));
+        assertTrue(booleanMatcher.isFailedPrimitive());
+
+        EqualFieldsMatcher<Byte> byteMatcher = Matchers.equalTo((byte) 42);
+        assertTrue(byteMatcher.matches((byte) 42));
+        assertFalse(byteMatcher.matches((byte) 24));
+        assertTrue(byteMatcher.isFailedPrimitive());
+
+        EqualFieldsMatcher<Character> charMatcher = Matchers.equalTo('a');
+        assertTrue(charMatcher.matches('a'));
+        assertFalse(charMatcher.matches('z'));
+        assertTrue(charMatcher.isFailedPrimitive());
+
+        EqualFieldsMatcher<Short> shortMatcher = Matchers.equalTo((short) 42);
+        assertTrue(shortMatcher.matches((short) 42));
+        assertFalse(shortMatcher.matches((short) 24));
+        assertTrue(shortMatcher.isFailedPrimitive());
+
+        EqualFieldsMatcher<Integer> integerMatcher = Matchers.equalTo(42);
+        assertTrue(integerMatcher.matches(42));
+        assertFalse(integerMatcher.matches(24));
+        assertTrue(integerMatcher.isFailedPrimitive());
+
+        EqualFieldsMatcher<Float> floatMatcher = Matchers.equalTo(42F);
+        assertTrue(floatMatcher.matches(42F));
+        assertFalse(floatMatcher.matches(24F));
+        assertTrue(floatMatcher.isFailedPrimitive());
+
+        EqualFieldsMatcher<Double> doubleMatcher = Matchers.equalTo(42.42);
+        assertTrue(doubleMatcher.matches(42.42));
+        assertFalse(doubleMatcher.matches(24.24));
+        assertTrue(doubleMatcher.isFailedPrimitive());
+
+        EqualFieldsMatcher<Long> longMatcher = Matchers.equalTo(42L);
+        assertTrue(longMatcher.matches(42L));
+        assertFalse(longMatcher.matches(24L));
+        assertTrue(longMatcher.isFailedPrimitive());
+
+        EqualFieldsMatcher<String> stringMatcher = Matchers.equalTo("foo");
+        assertTrue(stringMatcher.matches("foo"));
+        assertFalse(stringMatcher.matches("bar"));
+        assertTrue(stringMatcher.isFailedPrimitive());
     }
 }

--- a/test/src/test/java/org/axonframework/test/saga/CommandValidatorTest.java
+++ b/test/src/test/java/org/axonframework/test/saga/CommandValidatorTest.java
@@ -1,3 +1,19 @@
+/*
+ * Copyright (c) 2010-2022. Axon Framework
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package org.axonframework.test.saga;
 
 import org.axonframework.commandhandling.CommandMessage;
@@ -15,6 +31,11 @@ import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
 
+/**
+ * Test class validating the {@link CommandValidator}.
+ *
+ * @author Tom Soete
+ */
 class CommandValidatorTest {
 
     private CommandValidator testSubject;
@@ -48,6 +69,13 @@ class CommandValidatorTest {
         assertThrows(AxonAssertionError.class, () -> testSubject.assertDispatchedEqualTo(new SomeCommand("test")));
     }
 
+    @Test
+    void testMatchPrimitiveTypedCommands() {
+        when(commandBus.getDispatchedCommands()).thenReturn(listOfOneCommandMessage("some-string"));
+
+        assertThrows(AxonAssertionError.class, () -> testSubject.assertDispatchedEqualTo("some-other-string"));
+    }
+
     private List<CommandMessage<?>> emptyCommandMessageList() {
         return Collections.emptyList();
     }
@@ -56,8 +84,7 @@ class CommandValidatorTest {
         return Collections.singletonList(GenericCommandMessage.asCommandMessage(msg));
     }
 
-
-    private class SomeCommand {
+    private static class SomeCommand {
 
         private final Object value;
 

